### PR TITLE
Remove IVisualStudioWorkspaceHost and related interfaces

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This interface was originally introduced to support CodeLens, when we needed the ability to create a Workspace that was out-of-proc but connected to the in-process project system interfaces. This created the abstraction of how we pushed updates to the workspace, and came with the rule that the VisualStudioProjectTracker was unable to ever touch a workspace directly.

Today, that's not how we build our out-of-proc components. We now have an in-proc workspace, and we mirror that in-proc workspace via a pull model to the out-of-process workspace. This meant that the
interface was nearly pointless, except it was being used for the out-of-proc workspace setup to mirror some persistence settings over. Providing a new interface to do that directly is simpler and also
much cheaper, as it doesn't come with the overhead of pushing workspace changes to the remote host that were being ignored.

This has the direct perf benefit of removing some delays around setting up the remote workspace since we're not computing useless data. It has longer-term benefits to aid in further refactorings, as we now have workspace updates going directly to the in-proc workspace, which we can start to do in a free-threaded manner.

<details><summary>Ask Mode template</summary>

### Customer scenario

This is refactoring some ickier bits of our project system shims to enable us to make them free-threaded.

### Bugs this fixes

No particular bug, this is new feature work.

### Risk

Moderate: this is a tricky area of code with lots of tangled dependencies. (Trying to remove them is part of the point of this.)

### Performance impact

None expected, as this is largely removing layers of indirection.

</details>

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@sharwell, @heejaechang, @CyrusNajmabadi you might be interested in looking at the commit entitled "Reconnect the persistence service to working folder location changes" in this PR as it's the current refactoring I'm doing in the persistence service. There are still some TODOs and "DO NOT MERGE"s sprinkled in, but I think it makes the object lifetime a lot clearer and offers a nicely decoupled way for the IPersistencesStorageLocationService to expose location changes to the host.

I'd suggest you take a look but look at the Flight Level 350 view and see if I'm outright missing some sort of fundamental case this needed to support. There's still some special cases I know I lost and will try to clean up.

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+using Microsoft.CodeAnalysis.SolutionSize;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Storage
+{
+    /// <summary>
+    /// A service that enables storing and retrieving of information associated with solutions,
+    /// projects or documents across runtime sessions.
+    /// </summary>
+    internal abstract partial class AbstractPersistentStorageService : IPersistentStorageService2
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: I renamed the file to match; GitHub isn't necessarily picking up the rename.

File: src/Workspaces/Remote/ServiceHub/Services/RemoteHostService.cs [hunk 2]
```diff
@@ -261,14 +251,9 @@ private static TelemetrySession GetTelemetrySession(string serializedSession)
             return session;
         }
 
-        private static AbstractPersistentStorageService GetPersistentStorageService()
+        private static RemotePersistentStorageLocationService GetPersistentStorageService()
         {
-            // A bit slimy.  We just create an adhoc workspace so it will create the singleton
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/ServiceHub/Services/RemoteHostService.cs]: Note this picks up part of @heejaechang's change in #24458.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 3]
```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
         {
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: I guess projects given is already in right order to push?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 4]
```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
...
+
+            var projectInfos = inOrderToPush.Select(p => p.CreateProjectInfoForCurrentState()).ToImmutableArray();
+
+            // We need to enable projects to start pushing changes to the workspace even before we add the solution/project to the host.
+            // This is required because between the point we capture the project info for current state and the point where we start pushing to the workspace,
+            // project system may send new events on the AbstractProject on a background thread, and these won't get pushed over to the workspace hosts as we didn't set the _pushingChangesToWorkspaceHost flag on the AbstractProject.
+            // By invoking StartPushingToWorkspaceHosts upfront, any project state changes on the background thread will enqueue notifications to workspace hosts on foreground scheduled tasks.
+            foreach (var project in inOrderToPush)
+            {
+                project.StartPushingToWorkspaceHosts();
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: no WorkspaceHost anymore?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 5]
```diff
@@ -295,12 +333,11 @@ internal void RemoveProject(AbstractProject project)
 
             UpdateProjectBinPath(project, project.BinOutputPath, null);
 
-            using (Dispatcher.CurrentDispatcher.DisableProcessing())
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: not sure what make disable processing to removed and what make it to left out.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 6]
```diff
@@ -378,23 +415,17 @@ internal ProjectId GetOrCreateProjectIdForPath(string projectPath, string projec
...
+        /// Notifies the <see cref="Workspace"/> of the change with the appropriate threading handling.
         /// </summary>
         /// <remarks>This method must be called on the foreground thread.</remarks>
-        internal void NotifyWorkspaceHosts(Action<IVisualStudioWorkspaceHost> action)
+        internal void NotifyWorkspace(Action<Workspace> action)
         {
             AssertIsForeground();
 
             // We do not want to allow message pumping/reentrancy when processing project system changes.
             using (Dispatcher.CurrentDispatcher.DisableProcessing())
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: so this is to run action under dispatcher disable processing?

**Comment by CyrusNajmabadi** [comment]
Do i get a quid-pro-quo review in return @jasonmalinowski ? :)

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs [hunk 7]
```diff
@@ -5,19 +5,193 @@
...
+                    if (!temporary && !string.IsNullOrWhiteSpace(workingFolderPath))
+                    {
+                        OnWorkingFolderChanging_NoLock(
+                            new PersistentStorageLocationChangingEventArgs(
+                                visualStudioWorkspace.CurrentSolution.Id,
+                                workingFolderPath,
+                                mustUseNewStorageLocationImmediately: false));
+                    }
+                }
+                catch
```
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: NFW?

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs [hunk 8]
```diff
@@ -132,17 +132,8 @@ private enum GlobalNotificationState
...
-            // RegisterWorkspaceHost is required to be called from UI thread so push the code
-            // to UI thread to run. 
-            await Task.Factory.SafeStartNew(() =>
-            {
-                var projectTracker = vsWorkspace.GetProjectTrackerAndInitializeIfNecessary(Shell.ServiceProvider.GlobalProvider);
-
-                projectTracker.RegisterWorkspaceHost(host);
-                projectTracker.StartSendingEventsToWorkspaceHost(host);
-            }, CancellationToken.None, ForegroundThreadAffinitizedObject.CurrentForegroundThreadData.TaskScheduler).ConfigureAwait(false);
+            // TODO: subscribe to working folder changes
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: how are you intending to make this happen?

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs [hunk 9]
```diff
@@ -5,19 +5,193 @@
 using Microsoft.CodeAnalysis.Host;
 using Microsoft.CodeAnalysis;
 using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServices.Implementation.Extensions;
+using System;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: sort meeeeeeeeeee

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: can you undo the rename?

File: src/Workspaces/Core/Portable/Remote/IRemoteHostService.cs [hunk 10]
```diff
@@ -13,7 +13,6 @@ internal interface IRemoteHostService
         Task SynchronizeGlobalAssetsAsync(Checksum[] checksums, CancellationToken cancellationToken);
 
         void RegisterPrimarySolutionId(SolutionId solutionId, string storageLocation, CancellationToken cancellationToken);
-        void UnregisterPrimarySolutionId(SolutionId solutionId, bool synchronousShutdown, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Remote/IRemoteHostService.cs]: why?

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        private readonly IOptionService _optionService;
+        private readonly IPersistentStorageLocationService _locationService;
+        private readonly SolutionSizeTracker _solutionSizeTracker;
+
+        /// <summary>
+        /// This lock guards all mutable fields in this type.
+        /// </summary>
+        private readonly object _lock = new object();
+        private ReferenceCountedDisposable<IPersistentStorage> _currentPersistentStorage;
+        private SolutionId _currentPersistentStorageSolutionId;
```
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: so it cache the last solution rather than current primary solution.

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+                    // We do, great
+                    return PersistentStorageReferenceCountedDisposableWrapper.AddRefrenceCountToAndCreateWrapper(_currentPersistentStorage);
+                }
+
+                if (!SolutionSizeAboveThreshold(solution))
+                {
+                    return NoOpPersistentStorage.Instance;
+                }
+
+                string workingFolder = _locationService.TryGetStorageLocation(solution.Id);
```
Comment (heejaechang) [inline src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: so it no longer support case where solution is closed and new solution is opened, pending async task that hold onto last solution getting persistent service support. only new solution work.

one usage case before was that, when solution is closed but VS alive, pending task can keep saving its state to db so that it can be picked up next time. might not a bit deal with most people just kill VS. rather than close and reopen different solution.

File: src/Workspaces/Core/Portable/Remote/IRemoteHostService.cs (hunk 10 above)
Comment (heejaechang) [inline src/Workspaces/Core/Portable/Remote/IRemoteHostService.cs]: so this is basically send out to let it know current working folder.

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 8 above)
Comment (heejaechang) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: so, plan is just subscribe to location service's change event and sent that over in BG? since it is not tied to FG, that's probably fine.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Note this code is largely moving from VisualStudioWorkspaceTracker.WorkspaceHostState. If there's a better name I can use happy to fix it.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: No WorkspaceHost as a concept: that existed originally so we could support CodeLens prior to the Roslyn big switch. That time is long gone.

I should rename this method I guess?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: I'm somewhat convinced it doesn't do anything. This was added to prevent some reentrancy when we take locks...

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 6 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Yep. Whether it's useful I don't actually know.

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 8 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: Writing the next commit. I'll just have this subscribe to the in-proc IPersistentStorageLocationService and forward along.

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: Yes.

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs (hunk 11 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: Ah, right I can probably add a check that it's the one in the workspace.

Note: this is no longer a singleton but a proper workspace service, so there's one of these per workspace. Practically, there's only one SolutionId per workspace as well, although I do have a bug here thinking about it now where if you did have a change to the SolutionId and forgot to raise the storage location changed, I might fail to dispose the other one.

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs (hunk 12 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: @heejaechang Does this still work after your change in #24458? UnregisterPrimarySolution would get called once we closed down, and we'd dispose the storage and be done with it. There's no lifetime tracking we have for "somebody might still have a solution and they might ask for the storage sometime far later." And after somebody closed the solution they might have already moved it on disk, switched git branches, etc?

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: @CyrusNajmabadi I broke the rename out into a separate commit. Not great but will make things a bit easier I hope.

**Comment by jasonmalinowski** [comment]
> @CyrusNajmabadi: Do i get a quid-pro-quo review in return @jasonmalinowski ? :)

Sure, if you review this I'll review #24362. :-)

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 13]
```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
...
+
+                // Bail out if any of the following is true:
+                //  1. We have already started pushing changes for this project OR
+                //  2. We have already visited this project in a prior recursive call                
+                if (_pushedProjects.Contains(project) || !visited.Add(project))
+                {
+                    return;
+                }
+
+                foreach (var projectReference in project.GetCurrentProjectReferences())
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: ```GetCurrentProjectReferences```?  As opposed to what?  Previous project references?  Future references? (I'm not being snarky, i'm just tryign to figure out if there are special semantics here that i should be thinking about).

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 14]
```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
...
+            {
+                addToInOrderToPush(project);
+            }
+
+            var projectInfos = inOrderToPush.Select(p => p.CreateProjectInfoForCurrentState()).ToImmutableArray();
+
+            // We need to enable projects to start pushing changes to the workspace even before we add the solution/project to the host.
+            // This is required because between the point we capture the project info for current state and the point where we start pushing to the workspace,
+            // project system may send new events on the AbstractProject on a background thread, and these won't get pushed over to the workspace hosts as we didn't set the _pushingChangesToWorkspaceHost flag on the AbstractProject.
+            // By invoking StartPushingToWorkspaceHosts upfront, any project state changes on the background thread will enqueue notifications to workspace hosts on foreground scheduled tasks.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: wrap super long comments please.  100 or 120 columns seems good for github.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 15]
```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
...
+            foreach (var project in projects)
+            {
+                addToInOrderToPush(project);
+            }
+
+            var projectInfos = inOrderToPush.Select(p => p.CreateProjectInfoForCurrentState()).ToImmutableArray();
+
+            // We need to enable projects to start pushing changes to the workspace even before we add the solution/project to the host.
+            // This is required because between the point we capture the project info for current state and the point where we start pushing to the workspace,
+            // project system may send new events on the AbstractProject on a background thread, and these won't get pushed over to the workspace hosts as we didn't set the _pushingChangesToWorkspaceHost flag on the AbstractProject.
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: "the project system"

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: I would prefer that.  Maybe "StartProcessingWorkspaceChangeNotifications"?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 16]
```diff
@@ -272,11 +209,112 @@ internal void StartPushingToWorkspaceAndNotifyOfOpenDocuments(IEnumerable<Abstra
...
+                    var solutionInfo = SolutionInfo.Create(id, version.Value, solutionFilePath, projects: projectInfos);
+
+                    NotifyWorkspace(workspace => workspace.OnSolutionAdded(solutionInfo));
+
+                    _solutionAdded = true;
+
+                    var persistenceService = (VisualStudioPersistentStorageLocationService)WorkspaceServices.GetRequiredService<IPersistentStorageLocationService>();
+                    persistenceService.UpdateForVisualStudioWorkspace(_workspace);
+
+                }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: extra newline.

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: Ok.  let me know when that goes in.

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs [hunk 17]
```diff
@@ -5,19 +5,193 @@
...
     {
-        public bool IsSupported(Microsoft.CodeAnalysis.Workspace workspace)
+        private readonly IServiceProvider _serviceProvider;
+
+        private readonly object _gate = new object();
+        private SolutionId _currentSolutionId = null;
+        private string _currentWorkingFolderPath = null;
+        private bool _solutionEventsAdvised = false;
+
+        public event EventHandler<PersistentStorageLocationChangingEventArgs> StorageLocationChanging;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: can you give a threading comment on this?  like which threads this might fire on.

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs [hunk 18]
```diff
@@ -5,19 +5,193 @@
...
+                if (solutionId == _currentSolutionId)
+                {
+                    return _currentWorkingFolderPath;
+                }
+            }
+
+            return null;
+        }
+
+        internal void UpdateForVisualStudioWorkspace(Workspace visualStudioWorkspace)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: Maybe strongly type this to guarantee the right type?

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs [hunk 19]
```diff
@@ -5,19 +5,193 @@
...
+
+            StorageLocationChanging?.Invoke(this, eventArgs);
+
+            _currentSolutionId = eventArgs.SolutionId;
+            _currentWorkingFolderPath = eventArgs.NewStorageLocation;
+        }
+
+        private void ProcessChangeToIVsSolutionChange(Workspace visualStudioWorkspace)
+        {
+            lock (_gate)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: threading assert?

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs [hunk 20]
```diff
@@ -5,19 +5,193 @@
...
+
+            public EventSink(VisualStudioPersistentStorageLocationService service, Workspace visualStudioWorkspace)
+            {
+                _service = service;
+                _visualStudioWorkspace = visualStudioWorkspace;
+            }
+
+            void IVsSolutionWorkingFoldersEvents.OnQueryLocationChange(uint location, out bool pfCanMoveContent)
+            {
+                pfCanMoveContent = true;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: threading asserts for all of these methods?

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs (hunk 19 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: i would think this could only be called from the UI thread given htat it's called from a VS api sink below..  If that's the case, do we need the lock, or is all of this UI thread affinitized?

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: still seems awful on github.

File: src/Workspaces/Core/Portable/Remote/IRemoteHostService.cs (hunk 10 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Remote/IRemoteHostService.cs]: Right... but why did we get rid of Unregister.  Is that not needed anymore by the remote workspace?

File: src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/AbstractPersistentStorage.cs [hunk 21]
```diff
@@ -11,77 +11,34 @@ namespace Microsoft.CodeAnalysis.Host
 {
     internal abstract class AbstractPersistentStorage : IPersistentStorage
     {
-        private readonly IOptionService _optionService;
-        private readonly Action<AbstractPersistentStorage> _disposer;
-
-        private int _refCounter;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/AbstractPersistentStorage.cs]: Why is this no longer refcounted?

**Comment by CyrusNajmabadi** [comment]
Overall, looks good.  Would like the rename to be undone as it's harder to tell what's going on this PR.

File: src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/AbstractPersistentStorage.cs (hunk 21 above)
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/AbstractPersistentStorage.cs]: Ref counting is being done by the service directly via a wrapper pattern rather than requiring each implementation to do it.

This code was pretty tightly coupled in that the storage object was refcounting directly. The single instance was disposed from potentially multiple consumers (ick), and that _disposer called into the service, which called back into the object to actually manipulate ref counts, The types were pretty impressively coupled as a result.

Now, Dispose() means this actually goes away. For multiple consumers, we hand out a wrapper type which forwards to the single instance, and each Dispose is tracked nicely with @sharwell's type.

In other words: this type now has sane lifetime semantics in the absence of a service managing it. I'm not sure what it was before.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi You can look at 08d4ab6116b048269be064e79194a82ca4a1bf51 for without the rename.

**Comment by CyrusNajmabadi** [comment]
Wll do!

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs (hunk 7 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: @heejaechang It was previously this way. Thoughts?

Comment (heejaechang) [reply src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: at the time this was written, we didn't have NFW. now we have, so I think it is okay to capture NFW? 

NFW will only be an UI block issue when someone actually asking for full dump, otherwise, it is most likely not noticeable since all it does is sending out some telemetry.

File: src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs (hunk 12 above)
Comment (heejaechang) [reply src/Workspaces/Core/Desktop/Workspace/Storage/AbstractPersistentStorageService.cs]: I think we talked about this offline?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 14 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: This code is being moved, and deleted in another refactoring or two.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 13 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: This code is being moved, and deleted in another refactoring or two.

**Comment by jasonmalinowski** [comment]
@dotnet/roslyn-ide: OK, no longer WIP, and available for review.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs [hunk 22]
```diff
@@ -325,13 +325,15 @@ protected void SetIntellisenseBuildResultAndNotifyWorkspaceHosts(bool succeeded)
                 {
                     m[ProjectGuidPropertyName] = Guid;
                     m[nameof(LastDesignTimeBuildSucceeded)] = LastDesignTimeBuildSucceeded;
-                    m[nameof(PushingChangesToWorkspaceHosts)] = PushingChangesToWorkspaceHosts;
+
+                    // Use old name for consistency
+                    m["PushingChangesToWorkspaceHosts"] = PushingChangesToWorkspace;
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: I *just* added this too!

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs [hunk 23]
```diff
@@ -270,6 +244,52 @@ private void OnGlobalOperationStopped(object sender, GlobalOperationEventArgs e)
...
+        private void EnqueueStorageLocationChange(SolutionId solutionId, string storageLocation)
+        {
+            lock (_currentRemoteWorkspaceNotificationTaskGate)
+            {
+                _currentRemoteWorkspaceNotificationTask = _currentRemoteWorkspaceNotificationTask.SafeContinueWithFromAsync(_ =>
+                {
+                    return _rpc.InvokeAsync(
+                        nameof(IRemoteHostService.UpdateSolutionStorageLocation),
+                        new object[] { solutionId, storageLocation });
+                }, CancellationToken.None, TaskScheduler.Default);
```
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: Should we support cancellation here so the wait in Unregister can be shorter?

**Comment by CyrusNajmabadi** [comment]
I'll review later tonight.

File: src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs [hunk 24]
```diff
@@ -29,9 +32,11 @@ private enum Size
         private readonly Encoding _encoding = Encoding.UTF8;
         internal readonly IOptionService _persistentEnabledOptionService = new OptionServiceMock(new Dictionary<IOption, object>
         {
-            { PersistentStorageOptions.Enabled, true }
+            { PersistentStorageOptions.Enabled, true },
+            { StorageOptions.SolutionSizeThreshold, 100 }
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs]: 100 whats?!?!?!  what's the unit?!?!

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs (hunk 22 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/AbstractProject.cs]: Hence my annoyance when you merged this last week. :smile:

File: src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs (hunk 24 above)
Comment (jasonmalinowski) [reply src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs]: I think "solution size units", which are currently implemented as bytes of source code? https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Desktop/Workspace/SolutionSize/SolutionSizeTracker.cs

(@heejaechang might want to chime in here.)

**Comment by jasonmalinowski** [comment]
@Pilchie Do I get ask mode approval from you or somebody else in this case?

File: src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs (hunk 23 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/Remote/ServiceHubRemoteHostClient.cs]: We want to ensure that register/unregister things are happening cleanly and in order. The suspicion is the wait, in practice, is effectively zero, and only happens in cases like renaming solutions.

**Comment by Pilchie** [comment]
I'll approve when Jenkins and reviews are happy.

File: src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs (hunk 19 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Storage/VisualStudioPersistentStorageLocationService.cs]: All mutation is affinitized, but TryGetStorageLocation() looks at two fields while reading so we don't want any strange half-written states going on there.

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi Is there further review commentary coming?

**Comment by CyrusNajmabadi** [comment]
Nope.  I'm good.  Things look fine.  And i assume if theyr'e not other stuff will break if obvious and terrible ways.

File: src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs (hunk 24 above)
Comment (heejaechang) [reply src/VisualStudio/CSharp/Test/PersistentStorage/AbstractPersistentStorageTests.cs]: yes. it is sum of length of sourceText. so not exactly bytes since for unicode, 2 bytes is 1 length. if source text's encoding is ascii code, it will be 1 byte for 1 length. 

this is not intended to be precise. rough estimate is all we need. we first tried to read file size to make it bytes, which show up at RPS file read, moved to tree length which must already there when solution crawler is called.

**Comment by jasonmalinowski** [comment]
@dotnet-bot retest windows_release_vs-integration_prtest please.

**Comment by jasonmalinowski** [comment]
(starting a conversation on that failure internally, not sure if we have a dupe)

**Comment by jasonmalinowski** [comment]
@Pilchie again for ask mode. Tests are green, and @CyrusNajmabadi has also seemingly signed off, albeit neither with the review command nor a :+1:, which is out of character. :-)

**Comment by Pilchie** [comment]
Approved for 15.7 P2.  Thanks!

**Comment by heejaechang** [comment]
awesome!
